sw/device/lib/dif/dif_plic.h

```diff
@@ -192,8 +192,15 @@ dif_plic_result_t dif_plic_irq_pending_status_get(const dif_plic_t *plic,
  * Claims an IRQ and gets the information about the source.
  *
  * Claims an IRQ and returns the IRQ related data to the caller. This function
- * reads a target specific CC register. dif_plic_irq_complete must be called
- * after the claimed IRQ has been serviced.
+ * reads a target specific CC register. #dif_plic_irq_complete must be called
+ * in order to allow another interrupt with the same source id to be delivered.
+ * This usually would be done once the interrupt has been serviced.
+ *
+ * Other interrupts can arrive via the PLIC before a claim has been completed,
+ * but an interrupt with the same interrupt id will not be fired before the
+ * first is completed. This allows for nested interrupt handling, if needed.
+ *
+ * @see #dif_plic_irq_complete
```

Comment:
I think there are two parts to this:

- Ibex:
https://ibex-core.readthedocs.io/en/latest/exception_interrupts.html#nested-interrupt-exception-handling - which is not "supported" by default, but software can essentially make it happen from within an interrupt handler.

- PLIC:
> Other interrupts can arrive via the PLIC before a claim has been completed

It would be nice to have a reference to the corresponding place in the RISC-V spec? For example, I don't think that was the case (unless I have misread it) in the old RISC-V PLIC specification (before it has been removed from the privileged specification):
https://riscv.org//wp-content/uploads/2017/05/riscv-privileged-v1.10.pdf , sections "7.4 Interrupt Gateways", "7.10 Interrupt Claims", "7.11 Interrupt Completion", "7.12 Interrupt Flow".

The most recent RISC-V PLIC specification at:
https://github.com/riscv/riscv-plic-spec/blob/master/riscv-plic.adoc
Is much less detailed than the old spec. Possibly I missed it, but I couldn't map the `dif_plic_irq_complete` new description to the new spec either.

Could you please sanity me, and add a reference to the spec if you find relevant documentation.

Reply:
I'm not going to introduce a reference to the Ibex docs in the docs for this DIF, given we want this code to go with the peripheral itself. I've got some updated text about how claim/complete might work, which says they are compatible with nested interrupts, and how claim/complete can or cannot be nested.

I've added a reference to the PLIC spec at the top of the document though, which I do agree is useful. Hopefully that and some better text help to explain the (quite minimal) requirements on how claim/complete should be called.

Reply:
I do think that the ["Interrupt Completion"](https://github.com/riscv/riscv-plic-spec/blob/master/riscv-plic.adoc#interrupt-completion) header in the spec does have enough detail to tell software what to do, and indeed does not restrict how interrupts are nested etc (which is always the core's problem).

Reply:
What I meant was that the description implies that PLIC gateway will forward the interrupt to the PLIC core when a higher priority interrupt than the claimed one arrives. Meaning that even if the previous interrupt has not "completed", the PLIC core could assert EIP again.

I don't think that it was the case for the old PLIC spec. **I suppose the new spec is vague enough that it allows us to implement it this way?** But if this is the case,  maybe it is worth mentioning it.

Reply:
Oh I see what you mean, yeah I'll add something about `mip.MEIP` potentially getting re-asserted, but note the interrupt handler won't be invoked unless `mie.MEIE` is enabled - which inside an interrupt handler, has to be explicitly re-set.

Reply:
We discussed this offline, after I became hesitant about documenting how the `irq_o` lines mapped to a hart's `mie.*EIP` values. I'm not sure we want to make strong statements there, as it's up to the top level how those are hooked up, and certainly the DIF makes no change there.